ethcore/light/src/on_demand/mod.rs

```diff
@@ -0,0 +1,520 @@
+// Copyright 2015, 2016 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! On-demand chain requests over LES. This is a major building block for RPCs.
+//! The request service is implemented using Futures. Higher level request handlers
+//! will take the raw data received here and extract meaningful results from it.
+
+use std::collections::HashMap;
+
+use ethcore::basic_account::BasicAccount;
+use ethcore::encoded;
+use ethcore::receipt::Receipt;
+
+use futures::{Async, Poll, Future};
+use futures::sync::oneshot;
+use network::PeerId;
+
+use net::{Handler, Status, Capabilities, Announcement, EventContext, BasicContext, ReqId};
+use util::{Bytes, RwLock};
+use types::les_request::{self as les_request, Request as LesRequest};
+
+pub mod request;
+
+/// Errors which can occur while trying to fulfill a request.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Error {
+	/// Request was canceled.
+	Canceled,
+	/// No suitable peers available to serve the request.
+	NoPeersAvailable,
+	/// Request timed out.
+	TimedOut,
+}
+
+impl From<oneshot::Canceled> for Error {
+	fn from(_: oneshot::Canceled) -> Self {
+		Error::Canceled
+	}
+}
+
+/// Future which awaits a response to an on-demand request.
+pub struct Response<T>(oneshot::Receiver<Result<T, Error>>);
+
+impl<T> Future for Response<T> {
+	type Item = T;
+	type Error = Error;
+
+	fn poll(&mut self) -> Poll<T, Error> {
+		match self.0.poll().map_err(Into::into) {
+			Ok(Async::Ready(val)) => val.map(Async::Ready),
+			Ok(Async::NotReady) => Ok(Async::NotReady),
+			Err(e) => Err(e),
+		}
+	}
+}
+
+type Sender<T> = oneshot::Sender<Result<T, Error>>;
+
+// relevant peer info.
+struct Peer {
+	status: Status,
+	capabilities: Capabilities,
+}
+
+// Attempted request info and sender to put received value.
+enum Pending {
```

Comment:
In the future, it might also carry around a set of peers who already failed to serve the request.